Import custom case counts

whiver opened the pull request:
## Related issues and PRs
- Contributes to #11 

## Description
The goal is to allow the user to import a custom data file instead of the predefined data in the "Confirmed cases" field.

This allows to compare simulated data with real world ones, specific to each organization.

## Impacted Areas in the application
Mainly the population card, where the "confirmed cases" dropdown now has a button to display the import popup.

The imported file is stored in the sessionStorage for now to avoid keeping the whole file in the state, and to keep the file between subsequent page refresh, without keeping it forever.

## Testing
- Create a tsv file with these values for example:

      time	cases	deaths	hospitalized	icu	recovered
4/24/2020	100	10	50	5	15

- Under the "Confirmed cases" field, click the import link and import your data
- Click "Run"

rneher commented:
Thanks @whiver. this doesn't seem to work atm for me, but the general direction is good I think. The main aim of the feature will be to allow individuals hospitals, counties, cities etc to upload their own case counts. It is therefore not so much the simulation output that we want to upload again, but custom case counts (the dots on the graph). The format of the files should be like those that we download though.

whiver commented:
> Thanks @whiver. this doesn't seem to work atm for me, but the general direction is good I think. The main aim of the feature will be to allow individuals hospitals, counties, cities etc to upload their own case counts. It is therefore not so much the simulation output that we want to upload again, but custom case counts (the dots on the graph). The format of the files should be like those that we download though.

Thanks for the confirmation :)
The current version is not working indeed, it's still a WIP, I only created the PR for visibility. I'm on it!

whiver commented:
> 
> 
> Thanks @whiver. this doesn't seem to work atm for me, but the general direction is good I think. The main aim of the feature will be to allow individuals hospitals, counties, cities etc to upload their own case counts. It is therefore not so much the simulation output that we want to upload again, but custom case counts (the dots on the graph). The format of the files should be like those that we download though.

Hi @rneher, **I now have a first working version** I got a couple of questions too:
- As we want to import the confirmed cases from the file, I guess that if a file is imported, we display the imported data instead of the default "Confirmed cases" data, right? I added a toggle button to switch between imported/default confirmed cases
- The exported file only contains the simulation results (displayed as lines in the graph) and not the confirmed cases (displayed as dots). This means that I had to perform a kind of mapping between the existing fields in the exported file and the scattered dots in the graph. More precisely:
     - `infectious` in the file becomes `Cumulative confirmed cases` in the graph
     - `fatality` in the file becomes `Cumulative confirmed deaths` in the graph
     - `severe` in the file becomes `Patients in hospital` in the graph
     - `ICU` in the file becomes `Patients in ICU` in the graph

     Is this kind of mapping what you expected? Of course we may add more dot types to the graph later, but I'll first start with the existing ones.
- I also updated the model because there was a mix between the exported data model and the internal one: `UserResult` (which seems to be export-related as it holds `ExportedTimePoint`) was using in `AlgorithmResult` too, so I created a dedicated interface named `SimulationResult` to be used internally (with the main difference that fields in the export interface are all optionals).

I still need to test it thoroughly, but I first wanted to make sure that I was heading to the right direction. Thanks in advance for the feedback :)

nnoll commented:
Hey @whiver,

This looks great! Thanks for your contribution. A few thoughts/feedback on the PR as written

1. I don't really like the "Import" case data button underneath the results card next to the export options. If we would want to make this obvious for what kind of import it is, the option should be as close to the case data dropdown (or replace it) as possible, which is located within the population card. If this isn't possible, somewhere near the scenario header would be my second preference.
2. The exported file doesn't contain confirmed cases as it's not an output of the simulation. I think including this data in our export would ultimately be confusing. The input data we utilize, formatted as a tsv, can be found within the data subdirectory. 
3. Some kind of mapping is to be expected as we have to match our simulated compartments (susceptible, infectious, ...) onto the way COVID-19 data is reported. Namely, we typically receive cumulative case & death counts and current counts of hospitalizations & icus (although this is more the exception than the rule). Thus the mapping is:
      (a) Cumulative confirmed cases -> no equivalent
      (b) Cases past 3 days -> infectious
      (c) Cumulative fatality (data) -> Cumulative fatality (model)
      (d) Hospitalizations -> Severe
      (e) Icu (data) -> ICU + Overflow
I would imagine for importing of user data, if a new cases / day field is given we would prefer to use this over our window estimate now. This would be the only additional field I could imagine allowing.

Let me know if anything is unclear!

whiver commented:
> Hey @whiver,
> 
> This looks great! Thanks for your contribution. A few thoughts/feedback on the PR as written
> 
>     1. I don't really like the "Import" case data button underneath the results card next to the export options. If we would want to make this obvious for what kind of import it is, the option should be as close to the case data dropdown (or replace it) as possible, which is located within the population card. If this isn't possible, somewhere near the scenario header would be my second preference.
> 
>     2. The exported file doesn't contain confirmed cases as it's not an output of the simulation. I think including this data in our export would ultimately be confusing. The input data we utilize, formatted as a tsv, can be found within the data subdirectory.
> 
>     3. Some kind of mapping is to be expected as we have to match our simulated compartments (susceptible, infectious, ...) onto the way COVID-19 data is reported. Namely, we typically receive cumulative case & death counts and current counts of hospitalizations & icus (although this is more the exception than the rule). Thus the mapping is:
>        (a) Cumulative confirmed cases -> no equivalent
>        (b) Cases past 3 days -> infectious
>        (c) Cumulative fatality (data) -> Cumulative fatality (model)
>        (d) Hospitalizations -> Severe
>        (e) Icu (data) -> ICU + Overflow
>        I would imagine for importing of user data, if a new cases / day field is given we would prefer to use this over our window estimate now. This would be the only additional field I could imagine allowing.
> 
> 
> Let me know if anything is unclear!

Hi @nnoll and thanks for these precisions!
Just to make sure: so the model of file we should be able to import are located in `data/case-counts`, right? So the headers are:

    time	cases	deaths	hospitalized	icu	recovered


I completely agree with you about the UI, I was waiting for a confirmation before working on adding it in the Population section.

Thanks :)

nnoll commented:
@whiver. Yes, those are the headers. You could also consider importing as a json that could be validated from schemas/CaseCounts.yml

whiver commented:
@nnoll Oh nice, thanks for the suggestion! I think I'll add the support of the JSON format too, but I'll do that in a following PR to avoid bloating this one.

whiver commented on src/components/Main/Scenario/ConfirmedCasesDataPicker.tsx:
```diff
@@ -0,0 +1,85 @@
+const caseCountOptions = caseCountsNames.map((country) => ({ value: country, label: country }))
+caseCountOptions.push({ value: NONE_COUNTRY_NAME, label: getI18n().t(NONE_COUNTRY_NAME) })
+
+export function ConfirmedCasesDataPicker(props: ConfirmedCasesDataPickerProps) {
```
N/A, it's a react component.

whiver commented:
@nnoll I updated the PR to move the import button to the population section. The UI is not awesome, but tweaking this part involves creating a custom form field, so I'd prefer to improve the design in a separate PR, except if you think it's  really needed.

I updated the import schema to follow existing cases data files from the data folder. You can test importing for example `data/case-counts/brazil/BRA-Acre.tsv`. I tried to make the import more resilient to errors.

I still need to find out why on the environment deployed by Now, when you import a file in the popup, its name/size are missing (it displays (NaNKb)), while in local env it works correctly. I would also add an explanation in the import popup about the expected import format.

Feel free to test and tell me what you think! :smiley:

ivan-aksamentov commented:
@whiver If you rebase on (or merge) master, with #591 you should be able to debug the production deployments in Chromium-based browsers as follows:
 - in dev tools' "Source" tab in sidebar, click "Filesystem", then "+" button. Add project's root directory. Then click "Allow" when asked to give access. You now should be able to set breakpoints and use `debug` statements, watch call stack and variables. Sometimes it gets wonky - restart the browser in this case.

whiver commented:
Thanks @ivan-aksamentov, I fixed the issue in prod. I don't exactly understand why, but `{[...uploadedFiles.values()].map(({ name, size }: File) =>` was returning undefined name & size in prod but not in debug.

It should be working now, tell me what do you think of it :)

ivan-aksamentov commented:
@whiver Hi William

Sorry for the delay. Have been busy on other stuff.
I have rebased it and am merging this. Thanks for the contribution!
